Tech Insights documentation (#881)
RoadieHQ/marketing-site: content/docs/tech-insights/data-sources/index.md

---
title: Data Sources
publishedDate: '2022-11-15'
description: Managing Data Sources.
---

## Introduction

> Data Sources periodically collect Facts about your services from third-party vendors or arbitrary APIs. The Facts collected by Data Sources are used for defining Checks. Roadie comes with a handful of built-in Data Sources with predefined Facts that you can use without any setup. You can also define your own Data Sources, which are will use APIs or files in a repository as a source of truth.

Reviewer: ```suggestion
Data Sources periodically collect Facts about your services from third-party vendors or arbitrary APIs. The Facts collected by Data Sources are used for defining Checks. Roadie comes with a handful of built-in Data Sources with predefined Facts that you can use without any setup. You can also define your own Data Sources, which can use external APIs or files in a repository as a source of truth.
```

To manage Data Sources, go to Tech Insights → Data Sources.

### Using Built-in Data Sources

> By default, Roadie includes 5 data sources for common vendors such as Datadog, Snyk, and GitHub. All built-in Data Sources have pre-defined Facts that you can use for defining your Checks.

Reviewer: ```suggestion
By default, Roadie includes 5 data sources for common vendors such as Datadog, Snyk, and GitHub. All built-in Data Sources have pre-defined Facts you can use for defining your Checks.
```

> To use built-in Data Sources, you don’t need to do any set up. Jump to the next section to learn more about Checks. Note that you can’t delete nor trigger manual data updates on built-in Data Sources.

Reviewer: 'Checks' as clickable link?

Reviewer: 'Note that you can’t delete nor trigger manual data updates on built-in Data Sources.' 
Feels out of place here. Maybe move to the section where the triggering is explained?

### Defining your own Data Source

> Roadie Tech Insights lets you define arbitrary Data Sources. Typically, custom Data Sources rely on pinging an API or reading a configuration file in a repository. When adding a Data Source, you’ll also specify which Facts that are extracted from said source.

Reviewer: ```suggestion
Roadie Tech Insights lets you define arbitrary Data Sources. Typically, custom Data Sources rely on calling an API or reading a configuration file in a repository. When adding a Data Source, you’ll also specify which Facts that are extracted from said source.
```

Reviewer: We should create a 'ping' data source which we can use as a built-in replacement for uptimerobot

Author: And allow it to send updates to Slack using real time alerts 😁

> Roadie Tech Insights lets you define arbitrary Data Sources. Typically, custom Data Sources rely on pinging an API or reading a configuration file in a repository. When adding a Data Source, you’ll also specify which Facts that are extracted from said source.

Reviewer: ```suggestion
Roadie Tech Insights lets you define arbitrary Data Sources. Typically, custom Data Sources rely on pinging an API or reading a configuration file in a repository. When adding a Data Source, you’ll also specify which Facts are extracted from said source.
```

To add a Data Source, click on the “Add Data Source” button in the Data Sources page.

![Add Data Source](./data-sources-overview.png)

To set up a Data Source, you’ll be prompted to define a Data Provider and define which fields will be extracted as Facts from the data. Let’s start overviewing the options available for setting up a Data Provider:

![Add Data Provider](./data-provider-step.png)

1 - You must specify a type for that new Data Source. The _HTTP_ type lets you connect to an external API to pull in data, while the _Component repository file_ type lets you extract data from a file path in the corresponding repository of a given Component in your Catalog. As an example, let’s use “_Component repository file”_.

2 - Set the path to the file you want to extract data from in repositories, starting from the root. This can be anything from JSON files to YAML files.

3 - Try out what would be the response when testing specific entity from the location you have provided. If you were to get the `package.json` from a `sample-service` component, the Data Source would get something like this:

![Dry run](./dry-run-result.png)

> Now that you have data, let’s define what Facts interests you. You’ll do this through the Fields Extraction.

Reviewer: ```suggestion
Now that you have data, let’s define what Facts interest you. You’ll do this through the Fields Extraction.
```

> Now that you have data, let’s define what Facts interests you. You’ll do this through the Fields Extraction.

Reviewer: ```suggestion
Now that you have data, let’s define what Facts interests you. You’ll do this through the Fields Extraction section.
```

![Fields extraction](./field-extraction.png)

> 4 - Choose a parser to extract a Fact from the data obtained before. \****For the type “*Component repository file” this* can be either JSON or RegEx parser type, while for *“HTTP”\* data provider type, only JSON is supported. \*\***Keep in mind, JSON type may be used only for JSON type of files.

Reviewer: I think we need to figure out how to capitalize the abbreviation of `regular expressions`. I'm not sure it should be like it currently is

Author: I will change it to Regex

JSON type of parser uses [JSONPath Plus](https://jsonpath-plus.github.io/JSONPath/docs/ts/) syntax to extract data from JSON, while RegEx type uses [ECMAScript syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) to extract data from text.

5 - If you’re using the JSON parser, specify a path from the root of the object. For example _“version”,_ or “_scripts.test”_. If you’re using the RegEx parser, specify a valid expression with a capture group if extracting values. Please note the RegEx does not need slashes at the start or end.

> An example of RegEx parser could be if we retrieved following result from location we have provided

Reviewer: ```suggestion
An example of RegEx parser could be if we retrieved following result from location we have provided:
```
This looks a bit funky when read. Maybe replace with something like 'Let's look how we would do it with an example. If we want to retrieve specific line from the following result:'

![Regex result](./dry-run-yaml-result.png)

> If we wanted to retrieve Node version we could write a following RegEx:

Reviewer: ```suggestion
If we wanted to retrieve Node version we could write the following RegEx:
```

![Field extraction Regex](./field-extraction-regex.png)

6 - Select the type of the parsed value.

> 7 - Check facts will run dry run upon data source and newly created check and let you know what would be the result if check was being run against entity you have provided as a test entity.

Reviewer: ```suggestion
7 - 'Check facts' button will run dry run upon data source and newly created check and let you know what would be the result if check was being run against entity you have provided as a test entity.
```

> If you are happy with the created data source and checks you can proceed and click ‘Add fact’.

Reviewer: This will add an additional fact row right? If they are happy, they don't have to click anything

Author: You are right, maybe I can clarify a bit more that this is just in case they want to add extra fact

---
title: Data Sources
publishedDate: '2022-11-15'
description: Managing Data Sources.
---

## Introduction

Data Sources periodically collect Facts about your services from third-party vendors or arbitrary APIs. The Facts collected by Data Sources are used for defining Checks. Roadie comes with a handful of built-in Data Sources with predefined Facts that you can use without any setup. You can also define your own Data Sources, which can use external APIs or files in a repository as a source of truth.

To manage Data Sources, go to Tech Insights → Data Sources.

### Using Built-in Data Sources

> By default, Roadie includes 5 Data Sources for common vendors such as Datadog, Snyk, and GitHub. All built-in Data Sources have pre-defined Facts you can use for defining your [Checks](../checks//index.md).

Reviewer 2: This number will get out of date very quickly. Please omit it.

> To use built-in Data Sources, you don’t need to do any set up. Jump to the next section to learn more about Checks.

Reviewer 2: Would be nice to link to the place they should jump to to make it easy for the reader.

### Defining your own Data Source

Roadie Tech Insights lets you define arbitrary Data Sources. Typically, custom Data Sources rely on calling an API or reading a configuration file in a repository. When adding a Data Source, you’ll also specify which Facts are extracted from said source.

To add a Data Source, click on the “Add Data Source” button in the Data Sources page.

![Add Data Source](./data-sources-overview.png)

> To set up a Data Source, you’ll be prompted to define a Data Provider and define which fields will be extracted as Facts from the data. Let’s start overviewing the options available for setting up a Data Provider:

Reviewer 2: You don't have to explain Name and Description in depth, but I would at least mention them so the user knows they're on the correct page.

![Add Data Provider](./data-provider-step.png)

> 1 - You must specify a type for that new Data Source. The _HTTP_ type lets you connect to an external API to pull in data, while the _Component repository file_ type lets you extract data from a file path in the corresponding repository of a given Component in your Catalog. As an example, let’s use “_Component repository file”_.

Reviewer 2: ![Screenshot 2022-11-16 at 13 43 11](https://user-images.githubusercontent.com/562403/202196475-0d99b4c7-37d7-4af4-a348-a25c968db4ca.png)

There's a space missing between HTTP and type.

Reviewer: akshually. It's the font that makes it look like that. There is a space

Reviewer 2: Weird. Could we add two spaces then?

2 - Set the path to the file you want to extract data from in repositories, starting from the root. This can be anything from JSON files to YAML files.

3 - Try out what would be the response when testing specific entity from the location you have provided. If you were to get the `package.json` from a `sample-service` component, the Data Source would get something like this:

![Dry run](./dry-run-result.png)

Now that you have data, let’s define what Facts interest you. You’ll do this through the Fields Extraction section.

![Fields extraction](./field-extraction.png)

> 4 - Choose a parser to extract a Fact from the data obtained before. For the type “Component repository file” this can be either JSON or Regex parser type, while for “HTTP” data provider type, only JSON is supported. Keep in mind, JSON type may be used only for JSON type of files.

Author: Yeah, it is a bit weird. I wanted to emphasise they can't use yaml files for example 😁

> 4 - Choose a parser to extract a Fact from the data obtained before. For the type “Component repository file” this can be either JSON or Regex parser type, while for “HTTP” data provider type, only JSON is supported. Keep in mind, JSON type may be used only for JSON type of files.

Reviewer 2: I would just say that then.

> Even though YAML is a superset of JSON, this parser will only work on JSON files.

> JSON type of parser uses [JSONPath Plus](https://jsonpath-plus.github.io/JSONPath/docs/ts/) syntax to extract data from JSON, while Regex type uses [ECMAScript syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) to extract data from text.

Reviewer 2: Love that we link to these docs. Is there a ticket to provide these links (with `target=_blank`) in the UI also?

Author: Not that I know, will check.

5 - If you’re using the JSON parser, specify a path from the root of the object. For example _“version”,_ or “_scripts.test”_. If you’re using the Regex parser, specify a valid expression with a capture group if extracting values. Please note the Regex does not need slashes at the start or end.

Let's look how we would do it with an example. If we want to retrieve specific line from the following result:

![Regex result](./dry-run-yaml-result.png)

If we wanted to retrieve Node version we could write the following Regex:

> ![Field extraction Regex](./field-extraction-Regex.png)

Reviewer 2: This image is not working.

![Screenshot 2022-11-16 at 13 46 56](https://user-images.githubusercontent.com/562403/202197272-c9cae311-9f54-4c86-b7ae-3c72ca91c243.png)

![Regex result](./dry-run-yaml-result.png)

If we wanted to retrieve Node version we could write the following Regex:

![Field extraction Regex](./field-extraction-regex.png)

6 - Select the type of the parsed value.

7 - 'Check facts' button will run dry run upon data source and newly created check and let you know what would be the result if check was being run against entity you have provided as a test entity.

If you wish to add more facts you can do so by clicking ‘Add fact’.

After successfully adding a fact you will be able to select kind and type of services to which data source should apply and save newly created Data source by clicking ‘Save’ button.

![Data Source Entity Filter](../data-sources//data-source-entity-filter.png)

You should be able to see the created Data Source in the overview screen. Newly created Data Sources have a refresh cycle set to 24hours, but you can modify this value in 'Edit' screen, as well as trigger an update manually from the kebab menu.

![Trigger update](../data-sources//trigger-update.png)

Note that you can’t trigger manual data updates on built-in Data Sources.

> Now that you created the Data Source, you can define [Checks](../checks/) for that Data Source.

Reviewer 2: This one seems to be missing a "Next steps" title.
